Add erase() and noErase()

stalgiag opened the pull request:
closes #3759 

I have an [early example in the editor here](https://editor.p5js.org/stalgiag/sketches/1WJC1MtvS). This opens up some neat opportunities for layering graphics, and exposing hidden HTML elements behind the canvas.

#### Currently:
  1. `erase()` activates a state in which all drawing (see 3. for exceptions) is subtractive from its destination canvas
  2. `noErase()` deactivates  that state
  3. `image()` and `background()` do not become erasers during the `erase()` state. This is consistent with `fill()` and prevents behavior that will rarely be desired (using the background or image as an eraser). If someone still wants that behavior they can use `blendMode(REMOVE)` which is added with this PR.
  4. In order to accomplish 3, the most recent change to the `globalCompositingOperation` is stored as `p5.Renderer2D._cachedBlendMode`
  5. The strength (or opacity) of the erasing is dictating by the current fill and stroke color
  6. **No docs/examples/tests yet**
  7. ~~No WEBGL erasing. This might be difficult due to the clearing style of WEBGL. I can give it a shot though.~~

#### Open Questions:
  1. Do we like this interface? ie `erase()`/`noErase()` as opposed to other options (see discussion in issue at top)
  2. Related to 5 above, should the strength of the erase be based on current fill, or should `erase()` accept a parameter of a number between 0 - 255?
  3. Currently `stroke()` also erases. Should we keep this behavior?

#### My opinions
  1. There may be use cases I am missing, but this combined with the option for `blendMode(REMOVE)` seems to work nicely.
  2. Requiring a parameter makes sense  to me(ala `fill()`) but this contradicts my answer to 3 below.
  3. I like that it is possible to erase with `stroke()`. This may create unexpected effects at first but is relatively intuitive and offers a lot of flexibility. Adding a parameter is complicated by this though.

stalgiag commented:
Working with WEBGL in the same way now. [Here is a nearly identical example](https://editor.p5js.org/stalgiag/sketches/aAKV2jFA5).

Note: that the only difference is that the erase fill is not semi-transparent, which I removed because of some odd default color blending that is unrelated to this PR. I will open another issue after this.

lmccart commented:
this is exciting! I read through it quickly and all seems on the right track. I will respond in the next couple days with some thoughts on the open questions.

limzykenneth commented:
This looks amazing! 

1. The current interface seems reasonable to me. The combination of `erase()`, `noErase()` and `blendMode(REMOVE)` will require a bit of explaining so I wonder if `blendMode(REMOVE)` can be merged with `erase()` into something like `erase(BACKGROUND)`? If not the current interface is still ok.

2. This kinda contradicts with my suggestion to 1 but I'm ok with its current behaviour as well. Just a wild idea, it would be pretty wild to be able to subtract color with `erase(r, g, b)` I think, though the semantic may be confusing then.

3. Ok with `stroke()` erasing. See 2 for color subtraction which applies to strokes as well.

lmccart commented:
1. I like these options. I hear @limzykenneth's point but think it may not be too much of an issue to keep both. I imagine the sets of methods would mostly be used in different cases.
2. I think a parameter makes most sense. I find it difficult to wrap my head around `fill()`, `stroke()` and transparency combined.
3. Seems fine to me. I didn't realize this was even possible.

stalgiag commented:
Okay this should be ready for review now.

I added parameters for erasing strength, docs, and unit tests. I also cleaned up the code and moved the functions down to the Renderer level.

lmccart commented on src/core/constants.js:
```diff
@@ -436,6 +436,12 @@ export const UP_ARROW = 38;
+ * @final
+ * @default destination-out
+ */
+export const REMOVE = 'destination-out';
```
one small thing. if we're adding a blendMode option here, can you also add it to this list in the blendMode() reference: https://github.com/processing/p5.js/blob/0.9.0/src/core/rendering.js#L236

stalgiag replied:
Good point thank you!

lmccart commented:
The different strengthFill and strengthStroke were unexpected for me. I would've thought, like an eraser, it would have the same strength across both. But it's also interesting to play with. I'm good to keep this and see what emerges. :) I think this is good to merge after we resolve the one comment I left.
